test/integration/metrics_service_integration_test.cc

```diff
@@ -0,0 +1,120 @@
+#include "common/common/version.h"
+#include "common/grpc/codec.h"
+#include "common/grpc/common.h"
+
+#include "test/integration/http_integration.h"
+
+#include "api/metrics_service.pb.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Stats {
+namespace Metrics {
+
+class MetricsServiceIntegrationTest : public HttpIntegrationTest,
+                                      public testing::TestWithParam<Network::Address::IpVersion> {
+public:
+  MetricsServiceIntegrationTest()
+      : HttpIntegrationTest(Http::CodecClient::Type::HTTP1, GetParam()) {}
+
+  void createUpstreams() override {
+    HttpIntegrationTest::createUpstreams();
+    fake_upstreams_.emplace_back(new FakeUpstream(0, FakeHttpConnection::Type::HTTP2, version_));
+  }
+
+  void initialize() override {
+    config_helper_.addConfigModifier([](envoy::api::v2::Bootstrap& bootstrap) {
+      auto* metrics_service_cluster = bootstrap.mutable_static_resources()->add_clusters();
+      metrics_service_cluster->MergeFrom(bootstrap.static_resources().clusters()[0]);
+      metrics_service_cluster->set_name("metrics_service");
+      metrics_service_cluster->mutable_http2_protocol_options();
+
+      auto* metrics_sink = bootstrap.add_stats_sinks();
+      metrics_sink->set_name("envoy.metrics_service");
+      envoy::api::v2::MetricsServiceConfig config;
+      config.mutable_grpc_service()->mutable_envoy_grpc()->set_cluster_name("metrics_service");
+      MessageUtil::jsonConvert(config, *metrics_sink->mutable_config());
+
+    });
+
+    HttpIntegrationTest::initialize();
+    fake_upstreams_[1]->set_allow_unexpected_disconnects(true);
+  }
+
+  void waitForMetricsServiceConnection() {
+    fake_metrics_service_connection_ = fake_upstreams_[1]->waitForHttpConnection(*dispatcher_);
+  }
+
+  void waitForMetricsStream() {
+    metrics_service_request_ = fake_metrics_service_connection_->waitForNewStream(*dispatcher_);
+  }
+
+  void waitForMetricsRequest() {
+    envoy::api::v2::StreamMetricsMessage request_msg;
+    metrics_service_request_->waitForGrpcMessage(*dispatcher_, request_msg);
+    EXPECT_STREQ("POST", metrics_service_request_->headers().Method()->value().c_str());
+    EXPECT_STREQ("/envoy.api.v2.MetricsService/StreamMetrics",
+                 metrics_service_request_->headers().Path()->value().c_str());
+    EXPECT_STREQ("application/grpc",
+                 metrics_service_request_->headers().ContentType()->value().c_str());
+    EXPECT_TRUE(request_msg.envoy_metrics_size() > 0);
+    const Protobuf::RepeatedPtrField<::io::prometheus::client::MetricFamily>& envoy_metrics =
+        request_msg.envoy_metrics();
+    bool known_counter_exists = false;
+    bool known_gauge_exists = false;
+    int known_counter_value = -1;
+    int known_gauge_value = -1;
+    for (::io::prometheus::client::MetricFamily metrics_family : envoy_metrics) {
```

Comment:
It might be slightly cleaner to convert this into a `std::unordered_map` or the like and then write the match logic.

Reply:
i thought about this earlier, but felt building another container may not required here. So did not do. let me know if you feel strongly about it. i can make that change.